java.lang.IllegalArgumentException when building the user agent string

Hi,

I've seen a few bug reports that when the SDK tries to build the user agent and send it to OkHttp in some situations it led to an IllegalArgumentException. 

First situation is on a LG D851 device:

````
Caused by: java.lang.IllegalArgumentException: Unexpected char 0x2122 at 60 in User-Agent value: contentful.java/7.6.2(Android Runtime 0.9) Linux/3.4.0-Xceed™-D851
        at okhttp3.Headers$Builder.checkNameAndValue(SourceFile:321)
        at okhttp3.Headers$Builder.add(SourceFile:271)
        at okhttp3.Request$Builder.addHeader(SourceFile:177)
        at com.contentful.java.cda.interceptor.HeaderInterceptor.intercept(SourceFile:38)
        at okhttp3.internal.http.RealInterceptorChain.proceed(SourceFile:92)
        at okhttp3.internal.http.RealInterceptorChain.proceed(SourceFile:67)
        at com.contentful.java.cda.interceptor.HeaderInterceptor.intercept(SourceFile:37)
        at okhttp3.internal.http.RealInterceptorChain.proceed(SourceFile:92)
        at okhttp3.internal.http.RealInterceptorChain.proceed(SourceFile:67)
        at okhttp3.RealCall.getResponseWithInterceptorChain(SourceFile:185)
        at okhttp3.RealCall.execute(SourceFile:69)
        at retrofit2.OkHttpCall.execute(SourceFile:180)
        at retrofit2.adapter.rxjava.CallExecuteOnSubscribe.call(SourceFile:40)
        at retrofit2.adapter.rxjava.CallExecuteOnSubscribe.call(SourceFile:24)
        at rx.Observable.unsafeSubscribe(SourceFile:10151)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:33)
        at rx.Observable.unsafeSubscribe(SourceFile:10151)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:33)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:30)
        at rx.Observable.unsafeSubscribe(SourceFile:10151)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:33)
        at rx.Observable.unsafeSubscribe(SourceFile:10151)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:33)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:30)
        at rx.Observable.unsafeSubscribe(SourceFile:10151)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:33)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:30)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:30)
        at rx.Observable.subscribe(SourceFile:10247)
        at rx.Observable.subscribe(SourceFile:10214)
        at rx.observables.BlockingObservable.blockForSingle(SourceFile:444)
        at rx.observables.BlockingObservable.first(SourceFile:167)
        at com.contentful.java.cda.FetchQuery.all(SourceFile:43)
````

Second situation is on a few devices set with Arabic languages like fa_IR, ar_EG, ar_AE and bn_BD:

````
Caused by: java.lang.IllegalArgumentException: Unexpected char 0x6f7 at 20 in X-Contentful-User-Agent value: sdk contentful.java/۷.۶.۲; platform java/۰.۹.۰; os Android/۶.۰.۱; 
        at okhttp3.Headers$Builder.checkNameAndValue(SourceFile:321)
        at okhttp3.Headers$Builder.add(SourceFile:271)
        at okhttp3.Request$Builder.addHeader(SourceFile:177)
        at com.contentful.java.cda.interceptor.HeaderInterceptor.intercept(SourceFile:38)
        at okhttp3.internal.http.RealInterceptorChain.proceed(SourceFile:92)
        at okhttp3.internal.http.RealInterceptorChain.proceed(SourceFile:67)
        at com.contentful.java.cda.interceptor.HeaderInterceptor.intercept(SourceFile:37)
        at okhttp3.internal.http.RealInterceptorChain.proceed(SourceFile:92)
        at okhttp3.internal.http.RealInterceptorChain.proceed(SourceFile:67)
        at com.contentful.java.cda.interceptor.HeaderInterceptor.intercept(SourceFile:37)
        at okhttp3.internal.http.RealInterceptorChain.proceed(SourceFile:92)
        at okhttp3.internal.http.RealInterceptorChain.proceed(SourceFile:67)
        at okhttp3.RealCall.getResponseWithInterceptorChain(SourceFile:185)
        at okhttp3.RealCall.execute(SourceFile:69)
        at retrofit2.OkHttpCall.execute(SourceFile:180)
        at retrofit2.adapter.rxjava.CallExecuteOnSubscribe.call(SourceFile:40)
        at retrofit2.adapter.rxjava.CallExecuteOnSubscribe.call(SourceFile:24)
        at rx.Observable.unsafeSubscribe(SourceFile:10151)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:33)
        at rx.Observable.unsafeSubscribe(SourceFile:10151)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:33)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:30)
        at rx.Observable.unsafeSubscribe(SourceFile:10151)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:33)
        at rx.Observable.unsafeSubscribe(SourceFile:10151)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:33)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:30)
        at rx.Observable.unsafeSubscribe(SourceFile:10151)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeMap.call(SourceFile:33)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:30)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:48)
        at rx.internal.operators.OnSubscribeLift.call(SourceFile:30)
        at rx.Observable.subscribe(SourceFile:10247)
        at rx.Observable.subscribe(SourceFile:10214)
        at rx.observables.BlockingObservable.blockForSingle(SourceFile:444)
        at rx.observables.BlockingObservable.first(SourceFile:167)
        at com.contentful.java.cda.FetchQuery.all(SourceFile:43)
````

Suggestion is to not rely on system properties when running on an Android environment since you can't control what values will come from there.

Thanks!

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,8 @@
 All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
+## Version [9.1.1] - (TBD)
+- Fix: numbers in http headers default back to english.
 
 ## Version [9.1.0] - (2018-02-13)
 - Add: Support for [incoming-links](https://www.contentful.com/developers/docs/references/content-delivery-api/#/reference/search-parameters/links-to-entry/query-entries/console/java) by @shm.

--- a/pom.xml
+++ b/pom.xml
@@ -177,11 +177,6 @@
       <plugin>
         <artifactId>maven-assembly-plugin</artifactId>
         <configuration>
-          <archive>
-            <manifest>
-              <mainClass>com.contentful.generator.Main</mainClass>
-            </manifest>
-          </archive>
           <descriptorRefs>
             <descriptorRef>jar-with-dependencies</descriptorRef>
           </descriptorRefs>

--- a/src/main/java/com/contentful/java/cda/interceptor/ContentfulUserAgentHeaderInterceptor.java
+++ b/src/main/java/com/contentful/java/cda/interceptor/ContentfulUserAgentHeaderInterceptor.java
@@ -3,11 +3,11 @@ package com.contentful.java.cda.interceptor;
 import com.contentful.java.cda.Platform;
 
 import java.util.LinkedHashMap;
+import java.util.Locale;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 import static java.lang.String.format;
-import static java.util.Locale.getDefault;
 import static java.util.regex.Pattern.compile;
 
 /**
@@ -56,13 +56,13 @@ public class ContentfulUserAgentHeaderInterceptor extends HeaderInterceptor {
      */
     public static class Version {
 
-      private static final String VERSION_REGEX = "^([0-9]+).([0-9]+).([0-9]+)(.*)?$";
+      private static final String VERSION_REGEX = "^(\\p{N}+).(\\p{N}+).(\\p{N}+)(.*)?$";
       private static final Pattern VERSION_PATTERN = compile(VERSION_REGEX);
       private static final int VERSION_PATTERN_GROUP_COUNT = 4;
       private static final int VERSION_PATTERN_GROUP_MAJOR = 1;
       private static final int VERSION_PATTERN_GROUP_MINOR = 2;
       private static final int VERSION_PATTERN_GROUP_PATCH = 3;
-      private static final String STABILITY_REGEX = "^([a-zA-Z]+[0-9]*).*";
+      private static final String STABILITY_REGEX = "^\\p{Alpha}+\\p{Alnum}*";
       private static final Pattern STABILITY_PATTERN = compile(STABILITY_REGEX);
 
       /**
@@ -86,8 +86,6 @@ public class ContentfulUserAgentHeaderInterceptor extends HeaderInterceptor {
         if (version == null || version.length() <= 0) {
           return null;
         }
-
-        version = removeNonAsciiCharacters(version);
 
         if (!version.contains(".")) {
           return null;
@@ -205,9 +203,9 @@ public class ContentfulUserAgentHeaderInterceptor extends HeaderInterceptor {
        */
       @Override public String toString() {
         if (stability == null) {
-          return format(getDefault(), "%d.%d.%d", major, minor, patch);
+          return format(Locale.ENGLISH, "%d.%d.%d", major, minor, patch);
         } else {
-          return format(getDefault(), "%d.%d.%d-%s", major, minor, patch, stability);
+          return format(Locale.ENGLISH, "%d.%d.%d-%s", major, minor, patch, stability);
         }
       }
     }
@@ -329,14 +327,14 @@ public class ContentfulUserAgentHeaderInterceptor extends HeaderInterceptor {
     public String toString() {
       if (getVersion() == null) {
         return format(
-            getDefault(),
+            Locale.ENGLISH,
             "%s %s; ",
             getIdentifier(),
             getName()
         );
       } else {
         return format(
-            getDefault(),
+            Locale.ENGLISH,
             "%s %s/%s; ",
             getIdentifier(),
             getName(),

--- a/src/test/java/com/contentful/java/cda/interceptor/ContentfulUserAgentHeaderInterceptorTest.java
+++ b/src/test/java/com/contentful/java/cda/interceptor/ContentfulUserAgentHeaderInterceptorTest.java
@@ -8,6 +8,7 @@ import org.junit.Test;
 
 import java.lang.reflect.Field;
 import java.lang.reflect.Modifier;
+import java.util.Locale;
 
 import static com.contentful.java.cda.interceptor.ContentfulUserAgentHeaderInterceptor.Section.OperatingSystem.Android;
 import static com.contentful.java.cda.interceptor.ContentfulUserAgentHeaderInterceptor.Section.OperatingSystem.Linux;
@@ -216,14 +217,20 @@ public class ContentfulUserAgentHeaderInterceptorTest {
 
   @Test
   public void testIgnoreNonAsciiCharactersInStability() {
-    assertThat(Version.parse("3.4.0-Xceed™-D851").toString()).isEqualTo("3.4.0-Xceed-D851");
+    assertThat(Version.parse("3.4.0-Xceed™-D851").toString()).isEqualTo("3.4.0-Xceed");
   }
 
   @Test
   public void testIgnoreNonAsciiCharactersInVersionNumber() {
-    assertThat(Version.parse("۷.۶.۲")).isNull();
-    assertThat(Version.parse("۰.۹.۰")).isNull();
-    assertThat(Version.parse("۶.۰.۱")).isNull();
+    final Locale aDefault = Locale.getDefault();
+    Locale.setDefault(new Locale("ar", "EG"));
+    try {
+      assertThat(Version.parse("۷.۶.۲").toString()).isEqualTo("7.6.2");
+      assertThat(Version.parse("۰.۹.۰").toString()).isEqualTo("0.9.0");
+      assertThat(Version.parse("۶.۰.۱").toString()).isEqualTo("6.0.1");
+    } finally {
+      Locale.setDefault(aDefault);
+    }
   }
 
   @Test
@@ -237,7 +244,7 @@ public class ContentfulUserAgentHeaderInterceptorTest {
     final String name = header.getName();
 
     assertThat(name).isEqualTo("X-Contentful-User-Agent");
-    assertThat(value).isEqualTo("app my-app; ");
+    assertThat(value).isEqualTo("app my-app/1.0.9; ");
 
   }
 }

